File: src/cpp/ee/apps_flyer/private/AppsFlyerBridge.cpp
//
//  AppsFlyerBridge.cpp
//  ee_x_mobile_apps_flyer
//
//  Created by eps on 6/26/18.
//

#include "ee/apps_flyer/private/AppsFlyerBridge.hpp"

#include <ee/core/ILogger.hpp>
#include <ee/core/IMessageBridge.hpp>
#include <ee/core/Utils.hpp>
#include <ee/nlohmann/json.hpp>

namespace ee {
namespace apps_flyer {
namespace {
// clang-format off
const std::string kPrefix   = "AppsFlyerBridge";
const auto kInitialize      = kPrefix + "Initialize";
const auto kStartTracking   = kPrefix + "StartTracking";
const auto kGetDeviceId     = kPrefix + "GetDeviceId";
const auto kSetDebugEnabled = kPrefix + "SetDebugEnabled";
const auto kSetStopTracking = kPrefix + "SetStopTracking";
const auto kTrackEvent      = kPrefix + "TrackEvent";
const auto kLogAdRevenue    = kPrefix + "LogAdRevenue";
const auto kOnPurchaseValidated    = kPrefix + "OnPurchaseValidated";
// clang-format on
} // namespace

using Self = Bridge;

Self::Bridge(IMessageBridge& bridge, ILogger& logger,
             const Destroyer& destroyer)
    : bridge_(bridge)
    , logger_(logger)
    , destroyer_(destroyer) {}

Self::~Bridge() = default;

void Self::destroy() {
    destroyer_();
}

void Self::initialize(const std::string& devKey, const std::string& iosAppId, const std::string& appIdentity) {
    if (initialized_) {
        return;
    }
    initialized_ = true;
    appIdentity_ = appIdentity;

    nlohmann::json json;
    json["devKey"] = devKey;
    json["iosAppId"] = iosAppId;
    bridge_.call(kInitialize, json.dump());
}

void Self::startTracking() {
    bridge_.call(kStartTracking);
}

std::string Self::getDeviceId() const {
    return bridge_.call(kGetDeviceId);
}

void Self::setDebugEnabled(bool enabled) {
    bridge_.call(kSetDebugEnabled, core::toString(enabled));
}

void Self::setStopTracking(bool enabled) {
    bridge_.call(kSetStopTracking, core::toString(enabled));
}

void Self::trackEvent(const std::string& name,
                      const std::map<std::string, std::string>& values) {
    nlohmann::json json;
    json["name"] = name;
    json["values"] = values;
    bridge_.call(kTrackEvent, json.dump());
}

void Self::logRevenue(const ILibraryAnalytics::AdRevenue &adRevenue) {
    if (!adRevenue.isValid()) {
        return;
    }

    std::string mediationNetwork;
    switch (adRevenue.mediationNetwork) {
        case AdNetwork::AdMob:
            mediationNetwork = "admob";
            break;
        case AdNetwork::AppLovin:
            mediationNetwork = "applovin";
            break;
        default:
            mediationNetwork = "others";
            break;
    }

    std::string adFormat;
    switch (adRevenue.adFormat) {
        case AdFormat::AppOpen:
            adFormat = "App Open";
            break;
        case AdFormat::Banner:
            adFormat = "Banner";
            break;
        case AdFormat::Interstitial:
            adFormat = "Interstitial";
            break;
        case AdFormat::Rectangle:
            adFormat = "Rect";
            break;
        case AdFormat::Rewarded:
            adFormat = "Rewarded";
            break;
        case AdFormat::RewardedInterstitial:
            adFormat = "Rewarded Interstitial";
            break;
        default:
            adFormat = "Others";
            break;
    }

    nlohmann::json json;
    json["mediationNetwork"] = mediationNetwork;
    json["monetizationNetwork"] = adRevenue.monetizationNetwork;
    json["revenue"] = adRevenue.revenue;
    json["currencyCode"] = adRevenue.currencyCode;
    json["adFormat"] = adFormat;
    json["adUnit"] = adRevenue.adUnit;

    auto output = json.dump();
    logger_.info("%s: %s", kPrefix.c_str(), output.c_str());
    bridge_.call(kLogAdRevenue, output);
}

void Self::logRevenue(const ILibraryAnalytics::IapRevenue &iapRevenue) {
    // Prepare name
    std::string pIdShorten = iapRevenue.productId;
    if (!appIdentity_.empty()) {
        auto found = pIdShorten.find(appIdentity_ + ".");
        if (found != std::string::npos) {
            pIdShorten.erase(found, appIdentity_.length());
        }
    }

    {
        auto removeStr = "com.senspark.";
        auto found = pIdShorten.find(removeStr);
        if (found != std::string::npos) {
            pIdShorten.erase(found, std::string(removeStr).length());
        }
    }

    std::replace(pIdShorten.begin(), pIdShorten.end(), '.', '_');
    std::replace(pIdShorten.begin(), pIdShorten.end(), '-', '_');

    // Prepare Parameters
    auto evName = "conv_buy_iap_" + pIdShorten;

    nlohmann::json json;
    json["name"] = evName;
    json["values"] = {
        {"af_revenue", 0}, // Đã sử dụng thư viện track iap của AppsFlyer nên ko gửi price nữa
        {"af_currency", iapRevenue.currencyCode},
        {"af_quantity", 1},
        {"af_content_type", "iap_product"},
        {"af_content_id", iapRevenue.productId},
    };

    auto output = json.dump();

    logger_.info("%s %s: %s", kPrefix.c_str(), evName.c_str(), output.c_str());
    bridge_.call(kTrackEvent, output);
}
} // namespace apps_flyer
} // namespace ee

File: src/android/apps_flyer/src/main/java/com/ee/AppsFlyerBridge.kt
package com.ee

import android.app.Activity
import android.app.Application
import androidx.annotation.AnyThread
import com.appsflyer.AppsFlyerConversionListener
import com.appsflyer.AppsFlyerLib
import com.appsflyer.adrevenue.AppsFlyerAdRevenue
import com.appsflyer.adrevenue.adnetworks.generic.MediationNetwork
import com.appsflyer.adrevenue.adnetworks.generic.Scheme
import com.appsflyer.api.PurchaseClient
import com.appsflyer.api.Store
import com.appsflyer.internal.models.InAppPurchaseValidationResult
import com.appsflyer.internal.models.SubscriptionValidationResult
import com.ee.internal.deserialize
import com.ee.internal.serialize
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.longOrNull
import java.util.Currency
import java.util.Locale

class AppsFlyerBridge(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger,
    private val _application: Application,
    private var _activity: Activity?
) : IPlugin {
    companion object {
        private val kTag = AppsFlyerBridge::class.java.name
        private const val kPrefix = "AppsFlyerBridge"
        private const val kInitialize = "${kPrefix}Initialize"
        private const val kStartTracking = "${kPrefix}StartTracking"
        private const val kGetDeviceId = "${kPrefix}GetDeviceId"
        private const val kSetDebugEnabled = "${kPrefix}SetDebugEnabled"
        private const val kSetStopTracking = "${kPrefix}SetStopTracking"
        private const val kTrackEvent = "${kPrefix}TrackEvent"
        private const val kLogAdRevenue = "${kPrefix}LogAdRevenue"
        private const val kOnPurchaseValidated = "${kPrefix}OnPurchaseValidated"
    }

    private val _tracker = AppsFlyerLib.getInstance()

    init {
        _logger.info("$kTag: constructor begin: application = $_application activity = $_activity")
        registerHandlers()
        _logger.info("$kTag: constructor end")
    }

    override fun onCreate(activity: Activity) {
        _activity = activity
    }

    override fun onStart() {}
    override fun onStop() {}
    override fun onResume() {}
    override fun onPause() {}

    override fun onDestroy() {
        _activity = null
    }

    override fun destroy() {
        deregisterHandlers()
    }

    @Serializable
    private class InitializeRequest(
        val devKey: String,
        val iosAppId: String
    )

    @Serializable
    private class TrackEventRequest(
        val name: String,
        val values: Map<String, JsonPrimitive>
    )

    @Serializable
    @Suppress("unused")
    class AdRevenueData(
        val mediationNetwork: String,
        val monetizationNetwork: String,
        val currencyCode: String,
        val adFormat: String,
        val adUnit: String,
        val revenue: Double,
    )

    @Serializable
    @Suppress("unused")
    class PurchaseValidateData(
        val isSuccess: Boolean,
        val isTestPurchase: Boolean,
        val orderId: String,
        val productId: String,
    )

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerHandler(kInitialize) { message ->
            val request = deserialize<InitializeRequest>(message)
            initialize(request.devKey)
            ""
        }
        _bridge.registerHandler(kStartTracking) {
            startTracking()
            ""
        }
        _bridge.registerHandler(kGetDeviceId) {
            deviceId
        }
        _bridge.registerHandler(kSetDebugEnabled) { message ->
            setDebugEnabled(Utils.toBoolean(message))
            ""
        }
        _bridge.registerHandler(kSetStopTracking) { message ->
            setStopTracking(Utils.toBoolean(message))
            ""
        }
        _bridge.registerHandler(kTrackEvent) { message ->
            val request = deserialize<TrackEventRequest>(message)
            trackEvent(request.name, request.values)
            ""
        }
        _bridge.registerHandler(kLogAdRevenue) { message ->
            val request = deserialize<AdRevenueData>(message)
            logAdRevenue(request)
            ""
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kStartTracking)
        _bridge.deregisterHandler(kGetDeviceId)
        _bridge.deregisterHandler(kSetDebugEnabled)
        _bridge.deregisterHandler(kSetStopTracking)
        _bridge.deregisterHandler(kTrackEvent)
    }

    @AnyThread
    fun initialize(devKey: String) {
        Thread.runOnMainThread {
            val listener = object : AppsFlyerConversionListener {
                override fun onConversionDataSuccess(conversionData: Map<String, Any>) {
                    for (key in conversionData.keys) {
                        _logger.debug("$kTag: ${this::onConversionDataSuccess.name}: $key = ${conversionData[key]}")
                    }
                }

                override fun onConversionDataFail(errorMessage: String) {
                    _logger.debug("$kTag: ${this::onConversionDataFail.name}: $errorMessage")
                }

                override fun onAppOpenAttribution(conversionData: Map<String, String>) {
                    for (key in conversionData.keys) {
                        _logger.debug("$kTag: ${this::onAppOpenAttribution.name}: $key = ${conversionData[key]}")
                    }
                }

                override fun onAttributionFailure(errorMessage: String) {
                    _logger.debug("$kTag: ${this::onAttributionFailure.name}: $errorMessage")
                }
            }
            _tracker.init(devKey, listener, _application)
            _tracker.anonymizeUser(false)

            // af_ad_revenue
            val afRevenueBuilder = AppsFlyerAdRevenue.Builder(_application)
            AppsFlyerAdRevenue.initialize(afRevenueBuilder.build())

            val onValidated = { data: PurchaseValidateData ->
                val message = data.serialize()
                _bridge.callCpp(kOnPurchaseValidated, message)
            }

            // af_purchase
            val builder = PurchaseClient.Builder(_application, Store.GOOGLE)
                .logSubscriptions(true)
                .autoLogInApps(true)
                .setSandbox(false)
                .setInAppValidationResultListener(AppsFlyerIapResultListener(_logger, onValidated))
                .setSubscriptionValidationResultListener(
                    AppsFlyerSubscriptionResultListener(
                        _logger,
                        onValidated
                    )
                )

            val afPurchaseClient = builder.build()
            afPurchaseClient.startObservingTransactions()
        }
    }

    @AnyThread
    fun startTracking() {
        Thread.runOnMainThread {
            _tracker.start(_application)
        }
    }

    val deviceId: String
        @AnyThread get() = _tracker.getAppsFlyerUID(_application)!!

    @AnyThread
    fun setDebugEnabled(enabled: Boolean) {
        Thread.runOnMainThread {
            _tracker.setDebugLog(enabled)
        }
    }

    @AnyThread
    fun setStopTracking(enabled: Boolean) {
        Thread.runOnMainThread {
            _tracker.stop(enabled, _application)
        }
    }

    @AnyThread
    fun trackEvent(name: String, values: Map<String, JsonPrimitive>) {
        Thread.runOnMainThread {
            val parsedValues = HashMap<String, Any>()
            for ((key, value) in values) {
                if (value.isString) {
                    parsedValues[key] = value.content
                    continue
                }
                val long = value.longOrNull
                if (long != null) {
                    parsedValues[key] = long
                    continue
                }
                val double = value.doubleOrNull
                if (double != null) {
                    parsedValues[key] = double
                    continue
                }
                assert(false)
            }
            _tracker.logEvent(_application, name, values)
        }
    }

    private fun logAdRevenue(revenueData: AdRevenueData) {
        val customParams: MutableMap<String, String> = HashMap()
        customParams[Scheme.AD_UNIT] = revenueData.adUnit
        customParams[Scheme.AD_TYPE] = revenueData.adFormat
        customParams[Scheme.PLACEMENT] = "place"
        customParams[Scheme.ECPM_PAYLOAD] = "encrypt"

        val mediationNetwork = when (revenueData.mediationNetwork) {
            "applovin" -> MediationNetwork.applovinmax
            "admob" -> MediationNetwork.googleadmob
            else -> MediationNetwork.customMediation
        }

        _logger.info("$kTag: ${this::logAdRevenue.name}: ${revenueData.mediationNetwork} ${revenueData.revenue}")
        AppsFlyerAdRevenue.logAdRevenue(
            revenueData.monetizationNetwork,
            mediationNetwork,
            Currency.getInstance(Locale.US),
            revenueData.revenue,
            customParams
        )
    }
}


class AppsFlyerIapResultListener(
    private val _logger: ILogger,
    private val _onValidated: (AppsFlyerBridge.PurchaseValidateData) -> Unit,
) : PurchaseClient.InAppPurchaseValidationResultListener {

    companion object {
        private val kTag = AppsFlyerIapResultListener::class.java.name
    }

    override fun onResponse(result: Map<String, InAppPurchaseValidationResult>?) {
        result?.forEach { (k: String, v: InAppPurchaseValidationResult?) ->
            if (v.success && v.productPurchase != null) {
                val productPurchase = v.productPurchase!!
                val orderId = productPurchase.orderId
                val isTest =
                    productPurchase.purchaseType != null && productPurchase.purchaseType == 0
                val productId = productPurchase.productId
                val isSuccess = productPurchase.purchaseState == 0

                log("Validation success: $k $orderId $isTest $productId")
//                log("Product info $productPurchase")
                _onValidated(
                    AppsFlyerBridge.PurchaseValidateData(
                        isSuccess,
                        isTest,
                        orderId,
                        productId
                    )
                )
            } else {
                val failureData = v.failureData
                log("Validation fail: $k $failureData")
            }
        }
    }

    override fun onFailure(result: String, error: Throwable?) {
        log("Validation fail: $result, $error");
    }

    private fun log(message: String) {
        _logger.info("$kTag: $message")
    }
}

class AppsFlyerSubscriptionResultListener(
    private val _logger: ILogger,
    private val _onValidated: (AppsFlyerBridge.PurchaseValidateData) -> Unit,
) : PurchaseClient.SubscriptionPurchaseValidationResultListener {

    companion object {
        private val kTag = AppsFlyerSubscriptionResultListener::class.java.name
    }

    override fun onResponse(result: Map<String, SubscriptionValidationResult>?) {
        result?.forEach { (k: String, v: SubscriptionValidationResult?) ->
            val productPurchase = v.subscriptionPurchase
            if (v.success && productPurchase != null) {
                val isSuccess = productPurchase.subscriptionState == "SUBSCRIPTION_STATE_ACTIVE"
                val isTest = productPurchase.testPurchase != null
                val orderId = productPurchase.latestOrderId
                val productId = productPurchase.lineItems[0].productId

//                log("Validation success: $k $productPurchase")
                log("Validation success: $k $orderId $isTest $productId")

                _onValidated(
                    AppsFlyerBridge.PurchaseValidateData(
                        isSuccess,
                        isTest,
                        orderId,
                        productId
                    )
                )
            } else {
                val failureData = v.failureData
                log("Validation fail: $k $failureData")
            }
        }
    }

    override fun onFailure(result: String, error: Throwable?) {
        log("Validation fail: $result, $error")
    }

    private fun log(message: String) {
        _logger.info("${kTag}: $message")
    }
}